question2/odbcdemo.c

```diff
@@ -0,0 +1,288 @@
+    }
+
+    rc =
+        SQLGetInfo(hdbc, SQL_DRIVER_VER, (SQLPOINTER)ver, sizeof(ver), &strLen);
+    if ((rc != SQL_SUCCESS) && (rc != SQL_SUCCESS_WITH_INFO)) {
+        printf("SQLGetInfo has Failed. RC=%d\n", rc);
+        ODBC_error(henv, hdbc, hstmt);
+    }
+
+    printf("Driver version: %s\n", ver);
+    printf("Calling SQLGetTypeInfo...\n");
+    rc = SQLGetTypeInfo((SQLHSTMT)hstmt, SQL_ALL_TYPES);
+    if ((rc != SQL_SUCCESS) && (rc != SQL_SUCCESS_WITH_INFO)) {
+        printf("SQLGetTypeInfo has Failed. RC=%d\n", rc);
+        ODBC_error(henv, hdbc, hstmt);
+        EnvClose(henv, hdbc);
+        exit(255);
+    }
+
+    rc = SQLBindCol(hstmt, 1, SQL_C_CHAR, &dataStruct[0].charCol1[0],
+                    (SDWORD)sizeof(dataStruct[0].charCol1),
+                    (SQLLEN *)&dataStruct[0].length1);
+    if ((rc != SQL_SUCCESS) && (rc != SQL_SUCCESS_WITH_INFO)) {
+        printf("SQLBindCol(1) has Failed. RC=%d\n", rc);
+        ODBC_error(henv, hdbc, hstmt);
+        EnvClose(henv, hdbc);
+        exit(255);
+    }
+
+    rc = SQLBindCol(hstmt, 4, SQL_C_CHAR, &dataStruct2[0].charCol1[0],
+                    (SDWORD)sizeof(dataStruct2[0].charCol1),
+                    (SQLLEN *)&dataStruct2[0].length1);
+    if ((rc != SQL_SUCCESS) && (rc != SQL_SUCCESS_WITH_INFO)) {
+        printf("SQLBindCol(4) has Failed. RC=%d\n", rc);
+        ODBC_error(henv, hdbc, hstmt);
+        EnvClose(henv, hdbc);
+        exit(255);
+    }
+
+    rc = SQLBindCol(hstmt, 5, SQL_C_CHAR, &dataStruct3[0].charCol1[0],
+                    (SDWORD)sizeof(dataStruct3[0].charCol1),
+                    (SQLLEN *)&dataStruct3[0].length1);
+    if ((rc != SQL_SUCCESS) && (rc != SQL_SUCCESS_WITH_INFO)) {
+        printf("SQLBindCol(5) has Failed. RC=%d\n", rc);
+        ODBC_error(henv, hdbc, hstmt);
+        EnvClose(henv, hdbc);
+        exit(255);
+    }
+
+    // Set to bogus
+    dataStruct[0].length1 = -1;
+    dataStruct2[0].length1 = -1;
+    dataStruct3[0].length1 = -1;
+
+    int count = 0;
+
+    printf("Fetching result set...\n");
+    while (SQLFetch(hstmt) == SQL_SUCCESS) {
+        char *ptr;
+
+        if (count == 0) {
+            printf("LITERAL PREFIX\n");
+            printf("indicator(d): %d\n", dataStruct2[0].length1);
+            printf("indicator(lld): %lld\n", dataStruct2[0].length1);
+            printf("indicator(hex): ");
+
+            /* assign to char so we can dump out each byte */
+            ptr = (char *)&dataStruct2[0].length1;
+
+            for (i = 0; i < 8; i++) {
+                printf("%x", ptr[i]);
+            }
+            printf("\n");
+            printf("%-32s\n\n", dataStruct2[0].length1 == SQL_NULL_DATA
+                                    ? (UCHAR *)"NULL"
+                                    : dataStruct2[0].charCol1);
+
+            printf("LITERAL SUFFIX\n");
+            printf("indicator(d): %d\n", dataStruct3[0].length1);
+            printf("indicator(lld): %lld\n", dataStruct3[0].length1);
```

Comment:
using the wrong format specifier will cause a compiler warning if -Wformat flag is set. if the data type of DataInfoStruct::length1 stays as a long, the format specifier used should be %ld
```suggestion
            printf("indicator(ld): %ld\n", dataStruct3[0].length1);
```
if it changes to SQLLEN, the first printf should be removed. since SQLLEN can be either 32 bit or 64 bit, it may also be necessary to upcast to 64 bit int
ref: https://learn.microsoft.com/en-us/sql/odbc/reference/odbc-64-bit-information?view=sql-server-ver17
```suggestion
            printf("indicator(lld): %lld\n", dataStruct3[0].length1);
```